### ISSUE TITLE
How catboost handle with big data?

### ISSUE THREAD

**Issue by Mechanix12** [catboost/catboost#283]
Hi!
I try to use catboost in kaggle competition. https://www.kaggle.com/c/talkingdata-adtracking-fraud-detection
The size of my train set about 40m rows with 14 features.
When i try to train model, kernel always dies without any errors...

**Comment by annaveronika** [comment]
How many categorical features are present? What is the output in stdout and stderr? What is the version of catboost you are using? Is it cpu or gpu? If gpu, then which one?

**Comment by annaveronika** [comment]
have you tried training on a small subset of this data?

**Comment by annaveronika** [comment]
If there are many categorical features with many different values it might lead to rise of model size and memory consumption. To control that you can do:
1. Provide used_ram_limit value to restrict memory usage
2. Set max_ctr_complexity to 1 or 2.
By default it's 4 and it means that you are looking on combination of 4 categorical features.
3. Set model_size_reg to a bigger value to penalize heavy combinations.

**Comment by Mechanix12** [comment]
I try with 4m train set and it works fine...
I have 6 categorical features, cpu, 0.6.3 version

**Comment by annaveronika** [comment]
Btw please update to 0.7

**Comment by Mechanix12** [comment]
I have tried all above, but nothing helped...

**Comment by Noxoomo** [comment]
Hi!

Could you run with max_ctr_complexity 1 and report results? 
What amount of RAM you have on system?
How many disk space available? 
Could you try to train cmd-line version?

**Comment by Mechanix12** [comment]
Hi!
I tried with max_ctr_complexity = 1 and get:
Aborted (core dumped)

I have 16gb ram and 16gb swap

**Comment by annaveronika** [comment]
What is the output in stdout and stderr?

**Comment by DictMiner** [comment]
100Gb+ with max_ctr_complexity = 1 and all systems go

**Comment by Noxoomo** [comment]
I've been able to succesufully run CatBoost with default settings and without two last features (2 timestamp fields)
For 184903890 lines CatBoost requires ≈100GB of CPU RAM
With --max-ctr-complexity 1 it'll need less, but still memory usage'll be pretty high

**Comment by Noxoomo** [comment]
And did you use Pool class to store data? If you use pandas DataFrame, for example, you'll have 2x memory consumption: one for pandas and one for internal CatBoost format

**Comment by annaveronika** [comment]
The plotting doesn't depend on data size, so yes, if training works, plotting should work also

**Comment by aparamon** [comment]
Memory usage is indeed the major limitation of CatBoost currently. Unlike stochastic gradient and neural network models, CatBoost requires all data to be readily available in memory for fast random sampling.

It seems however that CatBoost algorithm might be modified for batch training.
Similar to [--rsm parameter](https://tech.yandex.com/catboost/doc/dg/concepts/python-reference_parameters-list-docpage)
> Random subspace method. The percentage of features to use at each split selection, when features are selected over again at random.

is it possible to introduce random subspaces along rows? Think *rolling-tree* generator:
1. First N rows are read from pool,
2. M trees are generated,
3. First k<N rows are discarded, next k rows are read from pool,
4. goto 2.

With N configured by user, M and k are deducible from total number of rows and total number of iterations.
This should definitely work for random forests, and most likely for gradient-boosting tree models as well. What do you think?

**Comment by YulinGUO** [comment]
Hi,
   I'm doing the same competition in kaggle and trying to apply catboost.
With --max-ctr-complexity 1, I got still the error:

uncaught exception:
    address -> 0x7f0740001a58
    what() -> "cxxrt::bad_alloc"
    type -> std::bad_alloc
Aborted

CPU version, 100GB ram....
Plz help...

**Comment by LeData** [comment]
@Mechanix12, you need to go see the log of the kernel to get the error message. It doesn't show in the console. My money is on  "exit code 137" appearing there, stemming from the Kaggle docker reaching its memory limit and shutting down.
I am using it for the same competition and am struggling with the same issue. A 20 million row dataframe of ~300Mb with 5 cat features and 2 numerical ones will blow the 17GB of RAM after about 400 iterations.
I haven't found a way to do the training by batches to keep the memory usage in check. There may be ways to deal with the issue with used_ram_limit or the save_snapshot and snapshot_file features.

**Comment by Mechanix12** [comment]
@LeData It seems like big disadvantage of catboost, comparing with lightgbm and xgboost...
I wish the developers will do something with this problem :)

**Comment by LeData** [comment]
Upon further inspection and toying with used_ram_limit, it seems that getting the borders for the categorical variables is what uses lots of memory, not the training.
For example. I have ran a model that asked >15GB of memory for the cat features work and then used <4GB afterwards.

**Comment by DeepakLabh** [comment]
@LeData  Great - 
Do you know of any other methods of **batchwise training** other than this?

Thanks

**Comment by LeData** [comment]
From what I understand “baseline” is supposed to do that, but it doesn’t save the model per se, so I’m not sure about it’s efficiency.

**Comment by LeData** [comment]
@Noxoomo Actually, using Pools seems to use a lot of memory. For a 1Gb pandas dataframe without categorical features, defining a pool takes >4Gb of memory. I am unsure that the pool creation takes into account the data types of Pandas Dataframes.

**Comment by annaveronika** [comment]
@LeData @DeepakLabh @Mechanix12 @YulinGUO 
This issue has been here for a while. Let's discuss what is needed here. Does the latest version of catboost with max_ctr_complexity=1 work fine for you?

**Comment by annaveronika** [comment]
closing because of inactivity. Please create new issues, if there are any problems for you.

**Comment by agcala** [comment]
You may want to look at this [Stackoverflow answer](https://stackoverflow.com/questions/47019215/catboost-training-model-for-huge-data22gb-with-multiple-chunks)
